streams/shared/src/main/scala/zio/stream/ZSink.scala

```diff
@@ -1271,10 +1278,15 @@ object ZSink extends ZSinkPlatformSpecific {
       def extract(state: State) = state._2.await
     }
 
-    for {
-      _       <- assertPositive(units).toManaged_
-      current <- clock.currentTime(TimeUnit.NANOSECONDS).toManaged_
-      bucket  <- Ref.make((units, current)).toManaged_
+    def checkTokens(sum: Long, max: Long): Long = if (sum < 0) max else math.min(sum, max)
```

Comment:
Question: when `sum < 0`, should we subtract it from `max`? Otherwise when the bucket is depleted, there is always constant amount of `max` available in addition.

Reply:
You're right, I'll fix this.

Reply:
Actually, on second thought, the `checkTokens` function is called to guard against overflow. `sum` is computed exclusively from positive values, whose total value might exceed `Long.MaxValue` (which is what the if statement checks). If that happens, we want `max` tokens to be available, otherwise, if we are dealing with positive values (`sum` and `max`), just get the minimum value.